doc/user_guide/psy_data.rst

```diff
@@ -277,41 +277,120 @@ a detailed description) or any of the profiling wrapper libaries
 
 .. _psy_data_transformation:
 
-PSyData Transformation and ``PSyDataNode``
--------------------------------------------
-The base transformation to create the PSyData callbacks is
-``PSyDataTrans``:
+``PSyDataTrans``
+----------------
+Any transformation that uses the PSyData API works
+by inserting a special node into the PSyclone tree representation
+of the program. Only at program creation time is the
+actual code created that implements the API. The
+``PSyDataTrans`` transformation contained in
+``psyir/transformations/psy_data_trans.py`` is the base
+class for other transformations like profiling and
+kernel data extraction. All derived transformations
+mostly add specific validations, and provide
+parameters to ``PSyDataTrans``, including the class
+of the node to insert. After passing validation,
+``PSyDataTrans`` creates an instance of the class
+requested, and inserts it into the tree.
 
 .. autoclass:: psyclone.psyir.transformations.psy_data_trans.PSyDataTrans
     :members:
 
-It is very similar to the profile transformation and kernel extraction
-transformations. These transforms insert a special node into the AST,
-which at code creation time will generate the code to give access to
-the data fields. Both profile and kernel extraction
-nodes use ``PSyDataNode`` as a base class, which will create the 
-PSyData calls as described above. The difference is that these
-classes will provide different parameters to the ``PSyDataNode``,
-resulting in different code being created:
+``PSyDataNode``
```

Comment:
Hmm - OK, I can see that (I documented this in the user's guide under the assumption that a user might want to implement a helper library using the PSyData API, for which you don't need to know much of the developer's guide). Then some other stuff should also be moved, e.g. the profiling API description (ProfileInit etc). But considering that I have three or so additional PR waiting, and this could cause quite a few conflicts - can I just open a ticket? I am happy to work on this once the PSyData-classes stuff is done (which is around 50% done anyway). I might need some guidelines about auto-docs then, e.g. can we use auto-doc in the dev guide?

Reply:
Yes, opening a new ticket is OK. We do really all need to sit down and agree on what documentation goes where (and then write the rules down!).

Reply:
Done (#711). I am already working on this as part of #415.